File: packages/react-native-audio-api/ios/nodes/GainNode/GainNode.m
#import <GainNode.h>
#import "AudioContext.h"
#import "Constants.h"

@implementation GainNode

- (instancetype)initWithContext:(AudioContext *)context
{
  if (self = [super initWithContext:context]) {
    _gainParam = [[AudioParam alloc] initWithContext:context
                                               value:1
                                            minValue:-[Constants maxGain]
                                            maxValue:[Constants maxGain]];
    self.numberOfInputs = 1;
    self.numberOfOutputs = 1;
  }

  return self;
}

- (void)cleanup
{
  _gainParam = nil;
}

- (void)process:(AVAudioFrameCount)frameCount bufferList:(AudioBufferList *)bufferList;
{
  float time = [self.context getCurrentTime];
  float deltaTime = 1 / self.context.sampleRate;

  for (int frame = 0; frame < frameCount; frame += 1) {
    float currentGain = [_gainParam getValueAtTime:time];

    for (int bufferNum = 0; bufferNum < bufferList->mNumberBuffers; bufferNum += 1) {
      float *buffer = (float *)bufferList->mBuffers[bufferNum].mData;

      buffer[frame] *= currentGain;
    }

    time += deltaTime;
  }

  [super process:frameCount bufferList:bufferList];
}

@end

File: packages/react-native-audio-api/android/src/main/cpp/GainNode/GainNode.cpp
#include "GainNode.h"
#include "AudioContext.h"

namespace audioapi {

GainNode::GainNode(AudioContext *context) : AudioNode(context) {
  gainParam_ = std::make_shared<AudioParam>(context, 1.0, -MAX_GAIN, MAX_GAIN);
}

std::shared_ptr<AudioParam> GainNode::getGainParam() const {
  return gainParam_;
}

bool GainNode::processAudio(float *audioData, int32_t numFrames) {
  if (!AudioNode::processAudio(audioData, numFrames)) {
    return false;
  }

  auto time = context_->getCurrentTime();
  auto deltaTime = 1.0 / context_->getSampleRate();

  for (int i = 0; i < numFrames * channelCount_; i++) {
    audioData[i] *= gainParam_->getValueAtTime(time);
    time += deltaTime;
  }

  return true;
}

} // namespace audioapi
